src/osd_field.c

```diff
@@ -0,0 +1,359 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#include <assert.h>
+#include <wlr/util/log.h>
+#include "common/mem.h"
+#include "config/rcxml.h"
+#include "view.h"
+#include "workspaces.h"
+#include "labwc.h"
+#include "osd.h"
+
+/* includes '%', terminating 's' and NULL byte, 8 is enough for %-9999W */
+#define LAB_FIELD_SINGLE_FMT_MAX_LEN 8
+static_assert(LAB_FIELD_SINGLE_FMT_MAX_LEN <= 255, "fmt_len is a unsigned char");
+
+/* forward declares */
+typedef void field_set_fn_t(struct buf *buf, struct view *view, const char *format);
+struct field_handlers {
+	const char fmt_char;
+	field_set_fn_t *fn;
+};
+
+static const struct field_handlers field_handlers[];
+
+/* Internal helpers */
+
+static const char *
+get_app_id_or_class(struct view *view, bool trim)
+{
+	/*
+	 * XWayland clients return WM_CLASS for 'app_id' so we don't need a
+	 * special case for that here.
+	 */
+	const char *identifier = view_get_string_prop(view, "app_id");
+
+	/* remove the first two nodes of 'org.' strings */
+	if (trim && identifier && !strncmp(identifier, "org.", 4)) {
+		char *p = (char *)identifier + 4;
+		p = strchr(p, '.');
+		if (p) {
+			return ++p;
+		}
+	}
+	return identifier;
+}
+
+static const char *
+get_type(struct view *view, bool short_form)
+{
+	switch (view->type) {
+	case LAB_XDG_SHELL_VIEW:
+		return short_form ? "[W]" : "[xdg-shell]";
+#if HAVE_XWAYLAND
+	case LAB_XWAYLAND_VIEW:
+		return short_form ? "[X]" : "[xwayland]";
+#endif
+	}
+	return "???";
```

Comment:
Why the three question marks?
Should we assert here as we should never reach this point.

Reply:
3 because this function always returns three chars.
Agree on some `assert(false)`, we still need some return after the assert though in case labwc is compiled without asserts.